fastcrypto/src/encoding.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use base64ct::Encoding as _;
+use eyre::{eyre, Result};
+use schemars::JsonSchema;
+use serde;
+use serde::de::{Deserializer, Error};
+use serde::ser::Serializer;
+use serde::Deserialize;
+use serde::Serialize;
+use serde_with::{DeserializeAs, SerializeAs};
+use std::fmt::Debug;
+
+#[inline]
+fn to_custom_error<'de, D, E>(e: E) -> D::Error
+where
+    E: Debug,
+    D: Deserializer<'de>,
+{
+    Error::custom(format!("byte deserialization failed, cause by: {:?}", e))
+}
+
+pub trait Encoding {
+    fn decode(s: &str) -> Result<Vec<u8>>;
+    fn encode<T: AsRef<[u8]>>(data: T) -> String;
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq, JsonSchema)]
+#[serde(try_from = "String")]
+pub struct Base64(String);
+
+impl TryFrom<String> for Base64 {
+    type Error = eyre::Report;
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        // Make sure the value is valid base64 string.
+        Base64::decode(&value)?;
+        Ok(Self(value))
+    }
+}
+
+impl Base64 {
+    pub fn to_vec(&self) -> Result<Vec<u8>, eyre::Report> {
+        Self::decode(&self.0)
+    }
+
+    pub fn from_bytes(bytes: &[u8]) -> Self {
+        Self(Self::encode(bytes))
+    }
+
+    pub fn encoded(&self) -> String {
+        self.0.clone()
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, JsonSchema, Clone)]
+pub struct Hex(String);
+
+impl Hex {
+    #[cfg(test)]
+    pub fn from_string(s: &str) -> Self {
+        Hex(s.to_string())
+    }
+    pub fn to_vec(&self) -> Result<Vec<u8>, eyre::Report> {
+        Self::decode(&self.0)
+    }
+    pub fn from_bytes(bytes: &[u8]) -> Self {
+        Self(Self::encode(bytes))
+    }
+}
+
+impl Encoding for Hex {
+    fn decode(s: &str) -> Result<Vec<u8>, eyre::Report> {
+        decode_bytes_hex(s)
+    }
+
+    fn encode<T: AsRef<[u8]>>(data: T) -> String {
+        format!("0x{}", encode_bytes_hex(&data))
```

Comment:
this is a direct copy from Sui, is this a good behavior or too idiomatic?

Reply:
looks good to me